### PR TITLE
fixed issue "Remove-FinOpsHub fails when there's only one resource"

### PR DESCRIPTION
**Pull request by jamelachahbar**
### Get-HubIdentifier.ps1:
I optimized the logic to directly extract a valid 13-character alphanumeric string and improve efficiency. The current method generates substrings unnecessarily (tested it with Verbose Output), slowing down performance and producing irrelevant results.
Regex Matching: Uses a regular expression to directly extract 13-character alphanumeric strings, removing the need for substring generation.
Excludes Special Characters: Filters out matches with hyphens or underscores to ensure a clean identifier.
Efficiency: Processes each string once, significantly improving performance for large collections.

### Remove-FinOpsHub.ps1
This PR simplifies the Remove-FinOpsHub logic by replacing the .Reduce() method with a more sequential foreach loop for deleting resources. It improves error handling, ensuring each resource deletion attempt is properly logged and handled individually. The script now includes clearer verbose messages for filtered resources, deletion attempts, and success statuses. These changes provide better transparency and avoid potential issues caused by parallel deletion, making the script more predictable and easier to debug. 

Fixes # 
#554
### 🔬 How did you test this change?

> - [X] 🤏 Lint tests
> - [ ] 🤞 PS -WhatIf / az validate
> - [X] 👍 Manually deployed + verified
> - [X] 💪 Unit tests
> - [x] 🙌 Integration tests

### 🙋‍♀️ Do any of the following that apply?

> - [ ] 🚨 This is a breaking change.
> - [ ] 🤏 The change is less than 20 lines of code.

### 📑 Did you update `docs/changelog.md`?

> - [ ] ✅ Updated changelog (required for `dev` PRs)
> - [ ] ➡️ Will add log in a future PR (feature branch PRs only)
> - [ ] ❎ Log not needed (small/internal change)

### 📖 Did you update documentation?

> - [ ] ✅ Public docs in `docs` (required for `dev`)
> - [ ] ✅ Internal dev docs in `src` (required for `dev`)
> - [ ] ➡️ Will add docs in a future PR (feature branch PRs only)
> - [ ] ❎ Docs not needed (small/internal change)

### CODE REVIEW

File: src/powershell/Private/Get-HubIdentifier.ps1 [hunk 1]
```diff
@@ -12,33 +12,34 @@
 
     Returns the string '123456hyfpqje'.
 #>
-function Get-HubIdentifier
+function Get-HubIdentifier
```
Comment (flanakin) [inline src/powershell/Private/Get-HubIdentifier.ps1:15]: nit: No trailing spaces (linter warning)
```suggestion
function Get-HubIdentifier
```

File: src/powershell/Private/Get-HubIdentifier.ps1 [hunk 2]
```diff
@@ -12,33 +12,34 @@
 
     Returns the string '123456hyfpqje'.
 #>
-function Get-HubIdentifier
+function Get-HubIdentifier 
 {
-    param
-    (
+    param (
```
Comment (flanakin) [inline src/powershell/Private/Get-HubIdentifier.ps1:17]: Are you editing in VS Code with the recommended extensions? If so, can you confirm you don't have any personal preferences that are overriding the code style rules in the workspace? While we're totally open to changing code style, we want to keep it consistent to minimize unnecessary change back and forth between different styles. Most of the style rules should be applied automatically on save. 
```suggestion
    param
    (
```

File: src/powershell/Private/Get-HubIdentifier.ps1 [hunk 3]
```diff
@@ -12,33 +12,34 @@
...
+    $idPattern = '[a-z0-9]{13}'  # Matches exactly 13 alphanumeric characters
     $substrings = @()
-    foreach ($string in $Collection)
-    {
-        for ($startIndex = 0; $startIndex -lt $string.Length; $startIndex++)
-        {
-            for ($endIndex = 1; $endIndex -le ($string.Length - $startIndex); $endIndex++)
-            {
+
+    foreach ($string in $Collection) {
```
Comment (flanakin) [inline src/powershell/Private/Get-HubIdentifier.ps1:27]:
```suggestion
    foreach ($string in $Collection)
    {
```

File: src/powershell/Private/Get-HubIdentifier.ps1 [hunk 4]
```diff
@@ -12,33 +12,34 @@
...
+        # Use regex to find valid 13-character matches
+        $match = [regex]::Matches($string, $idPattern)
+        if ($match.Count -gt 0) {
```
Comment (flanakin) [inline src/powershell/Private/Get-HubIdentifier.ps1:30]:
```suggestion
        if ($match.Count -gt 0)
        {
```

File: src/powershell/Private/Get-HubIdentifier.ps1 [hunk 5]
```diff
@@ -12,33 +12,34 @@
...
+            $substrings += $match | ForEach-Object { $_.Value }
+        }
+
+        # Generate all possible substrings
+        for ($startIndex = 0; $startIndex -lt $string.Length; $startIndex++) {
```
Comment (flanakin) [inline src/powershell/Private/Get-HubIdentifier.ps1:35]:
```suggestion
        for ($startIndex = 0; $startIndex -lt $string.Length; $startIndex++)
        {
```

File: src/powershell/Private/Get-HubIdentifier.ps1 [hunk 6]
```diff
@@ -12,33 +12,34 @@
...
+            for ($endIndex = 1; $endIndex -le ($string.Length - $startIndex); $endIndex++) {
```
Comment (flanakin) [inline src/powershell/Private/Get-HubIdentifier.ps1:36]:
```suggestion
            for ($endIndex = 1; $endIndex -le ($string.Length - $startIndex); $endIndex++)
            {
```

File: src/powershell/Public/Remove-FinOpsHub.ps1 [hunk 7]
```diff
@@ -25,84 +25,91 @@
     .EXAMPLE
     Remove-FinOpsHub -Name MyHub -ResourceGroupName MyRG -KeepStorageAccount
 
-    Deletes a FinOps Hub named MyHub and deletes all associated resource except the storagea ccount.
+    Deletes a FinOps Hub named MyHub and deletes all associated resources except the storage account.
 #>
 
-function Remove-FinOpsHub
-{
+function Remove-FinOpsHub {
```
Comment (flanakin) [inline src/powershell/Public/Remove-FinOpsHub.ps1:31]:
```suggestion
function Remove-FinOpsHub
{
```

File: src/powershell/Public/Remove-FinOpsHub.ps1 [hunk 8]
```diff
@@ -25,84 +25,91 @@
...
     [CmdletBinding(SupportsShouldProcess, ConfirmImpact = 'High')]
-    param
-    (
+    param (
```
Comment (flanakin) [inline src/powershell/Public/Remove-FinOpsHub.ps1:33]:
```suggestion
    param
    (
```

File: src/powershell/Public/Remove-FinOpsHub.ps1 [hunk 9]
```diff
@@ -25,84 +25,91 @@
...
         [Parameter(Mandatory = $true, ParameterSetName = 'Name')]
-        [ValidateNotNullOrEmpty ()]
-        [string]
-        $Name,
+        [ValidateNotNullOrEmpty()]
+        [string]$Name,
```
Comment (flanakin) [inline src/powershell/Public/Remove-FinOpsHub.ps1:36]:
```suggestion
        [string]
        $Name,
```

File: src/powershell/Public/Remove-FinOpsHub.ps1 [hunk 10]
```diff
@@ -25,84 +25,91 @@
...
 
         [Parameter(ParameterSetName = 'Name')]
-        [string]
-        $ResourceGroupName,
+        [string]$ResourceGroupName,
```
Comment (flanakin) [inline src/powershell/Public/Remove-FinOpsHub.ps1:39]:
```suggestion
        [string]
        $ResourceGroupName,
```

File: src/powershell/Public/Remove-FinOpsHub.ps1 [hunk 11]
```diff
@@ -25,84 +25,91 @@
...
 
         [Parameter(Mandatory = $true, ParameterSetName = 'Object')]
-        [ValidateNotNullOrEmpty ()]
-        [psobject]
-        $InputObject,
+        [ValidateNotNullOrEmpty()]
+        [psobject]$InputObject,
```
Comment (flanakin) [inline src/powershell/Public/Remove-FinOpsHub.ps1:43]:
```suggestion
        [psobject]
        $InputObject,
```

File: src/powershell/Public/Remove-FinOpsHub.ps1 [hunk 12]
```diff
@@ -25,84 +25,91 @@
...
 
         [Parameter(ParameterSetName = 'Name')]
         [Parameter(ParameterSetName = 'Object')]
-        [switch]
-        $KeepStorageAccount,
+        [switch]$KeepStorageAccount,
```
Comment (flanakin) [inline src/powershell/Public/Remove-FinOpsHub.ps1:47]:
```suggestion
        [switch]
        $KeepStorageAccount,
```

File: src/powershell/Public/Remove-FinOpsHub.ps1 [hunk 13]
```diff
@@ -25,84 +25,91 @@
...
 
         [Parameter(ParameterSetName = 'Name')]
         [Parameter(ParameterSetName = 'Object')]
-        [switch]
-        $Force
+        [switch]$Force
```
Comment (flanakin) [inline src/powershell/Public/Remove-FinOpsHub.ps1:51]:
```suggestion
        [switch]
        $Force
```

File: src/powershell/Public/Remove-FinOpsHub.ps1 [hunk 14]
```diff
@@ -25,84 +25,91 @@
...
     )
 
     $context = Get-AzContext
-    if (-not $context)
-    {
+    if (-not $context) {
```
Comment (flanakin) [inline src/powershell/Public/Remove-FinOpsHub.ps1:55]:
```suggestion
    if (-not $context)
    {
```

File: src/powershell/Public/Remove-FinOpsHub.ps1 [hunk 15]
```diff
@@ -25,84 +25,91 @@
...
-    try
-    {
-
-        if ($PSCmdlet.ParameterSetName -eq 'Name')
-        {
-            if (-not [string]::IsNullOrEmpty($ResourceGroupName))
-            {
+    try {
+        if ($PSCmdlet.ParameterSetName -eq 'Name') {
+            if (-not [string]::IsNullOrEmpty($ResourceGroupName)) {
```
Comment (flanakin) [inline src/powershell/Public/Remove-FinOpsHub.ps1:61]:
```suggestion
    try
    {
        if ($PSCmdlet.ParameterSetName -eq 'Name')
        {
            if (-not [string]::IsNullOrEmpty($ResourceGroupName))
            {
```

File: src/powershell/Public/Remove-FinOpsHub.ps1 [hunk 16]
```diff
@@ -25,84 +25,91 @@
...
                 $hub = Get-FinOpsHub -Name $Name -ResourceGroupName $ResourceGroupName
-            }
-            else
-            {
+            } else {
```
Comment (flanakin) [inline src/powershell/Public/Remove-FinOpsHub.ps1:63]:
```suggestion
            }
            else
            {
```

File: src/powershell/Public/Remove-FinOpsHub.ps1 [hunk 17]
```diff
@@ -25,84 +25,91 @@
...
                 $hub = Get-FinOpsHub -Name $Name
                 $ResourceGroupName = $hub.Resources[0].ResourceGroupName
             }
-        }
-        else
-        {
+        } else {
```
Comment (flanakin) [inline src/powershell/Public/Remove-FinOpsHub.ps1:67]:
```suggestion
        }
        else
        {
```

File: src/powershell/Public/Remove-FinOpsHub.ps1 [hunk 18]
```diff
@@ -25,84 +25,91 @@
...
             $hub = $InputObject
             $Name = $hub.Name
             $ResourceGroupName = $hub.Resources[0].ResourceGroupName
         }
 
-        if (-not $hub)
-        {
+        if (-not $hub) {
```
Comment (flanakin) [inline src/powershell/Public/Remove-FinOpsHub.ps1:73]:
```suggestion
        if (-not $hub)
        {
```

File: src/powershell/Public/Remove-FinOpsHub.ps1 [hunk 19]
```diff
@@ -25,84 +25,91 @@
...
+        Write-Verbose -Message "Found FinOps Hub: $Name in resource group $ResourceGroupName"
+
         $uniqueId = Get-HubIdentifier -Collection $hub.Resources.Name
-        Write-Verbose -Message "Unique identifier: $uniqueId"
+        
+        $resources = Get-AzResource -ResourceGroupName $ResourceGroupName |
+            Where-Object -FilterScript { $_.Name -like "*$uniqueId*" -and ((-not $KeepStorageAccount) -or $_.ResourceType -ne "Microsoft.Storage/storageAccounts") }
+            Write-Verbose -Message "Filtered Resources: $($resources | ForEach-Object { $_.Name })"
+
+        if ($null -eq $resources) {
```
Comment (flanakin) [inline src/powershell/Public/Remove-FinOpsHub.ps1:85]:
```suggestion
        if ($null -eq $resources)
        {
```

File: src/powershell/Public/Remove-FinOpsHub.ps1 [hunk 20]
```diff
@@ -25,84 +25,91 @@
...
+            Write-Warning "No resources found to delete."
+            return $false
+        }
 
-        $resources = Get-AzResource -ResourceGroupName $ResourceGroupName | Where-Object -FilterScript { $_.Name -like "*$uniqueId*" -and ((-not $KeepStorageAccount) -or $_.ResourceType -ne "Microsoft.Storage/storageAccounts") }
+        Write-Verbose -Message "Resources to be deleted: $($resources | ForEach-Object { $_.Name })"
+
+        if ($PSCmdlet.ShouldProcess($Name, 'DeleteFinOpsHub')) {
```
Comment (flanakin) [inline src/powershell/Public/Remove-FinOpsHub.ps1:92]:
```suggestion
        if ($PSCmdlet.ShouldProcess($Name, 'DeleteFinOpsHub'))
        {
```

File: src/powershell/Public/Remove-FinOpsHub.ps1 [hunk 21]
```diff
@@ -25,84 +25,91 @@
...
+            $success = $true
+            foreach ($resource in $resources) {
+                try {
```
Comment (flanakin) [inline src/powershell/Public/Remove-FinOpsHub.ps1:95]:
```suggestion
            foreach ($resource in $resources)
            {
                try
                {
```

File: src/powershell/Public/Remove-FinOpsHub.ps1 [hunk 22]
```diff
@@ -25,84 +25,91 @@
...
+                    Write-Verbose -Message "Deleting resource: $($resource.Name)"
+                    Remove-AzResource -ResourceId $resource.ResourceId -Force:$Force -ErrorAction Stop
+                } catch {
```
Comment (flanakin) [inline src/powershell/Public/Remove-FinOpsHub.ps1:98]:
```suggestion
                }
                catch
                {
```

File: src/powershell/Public/Remove-FinOpsHub.ps1 [hunk 23]
```diff
@@ -25,84 +25,91 @@
...
+                    $success = $false
+                }
+            }
 
-        if ($PSCmdlet.ShouldProcess($Name, 'DeleteFinOpsHub'))
-        {
-            return ($resources | Remove-AzResource -Force:$Force).Reduce({ $args[0] -and $args[1] }, $true)            
+            return $success
+        }
+    } catch {
```
Comment (flanakin) [inline src/powershell/Public/Remove-FinOpsHub.ps1:106]:
```suggestion
    }
    catch
    {
```

File: src/powershell/Public/Remove-FinOpsHub.ps1 [hunk 24]
```diff
@@ -25,84 +25,91 @@
...
+        Write-Error -Message "Failed to remove FinOps hub. Error: $_"
+        if ($_.Exception.InnerException) {
+            throw "Detailed Error: $($_.Exception.InnerException.Message)"
+        } else {
```
Comment (flanakin) [inline src/powershell/Public/Remove-FinOpsHub.ps1:110]:
```suggestion
        if ($_.Exception.InnerException)
        {
            throw "Detailed Error: $($_.Exception.InnerException.Message)"
        }
        else
        {
```

**Review by flanakin — APPROVED** [review]
Everything looks good at a glance. My only concern would be not having tests to validate this. We're already missing tests for this command, so I'm not necessarily blocking on that. It's just part of why this wasn't caught in the first place 😕

File: src/powershell/Private/Get-HubIdentifier.ps1 (hunk 2 above)
Comment (jamelachahbar) [reply src/powershell/Private/Get-HubIdentifier.ps1:17]: I need to check that. Do we have a template so that next time I commit ps code it is not deviating from the current style? I don't have a personal preference. It is all good for me.

Comment (jamelachahbar) [reply src/powershell/Private/Get-HubIdentifier.ps1:17]: Do you want me to commit all suggestions per suggestion?

Comment (jamelachahbar) [reply src/powershell/Private/Get-HubIdentifier.ps1:17]: Yes. I need to check my settings. Do we have any guide or template? I am okay with whatever style we need to adhere to. Do you want me to commit the suggestions? I seenitbis per suggestion

Comment (flanakin) [reply src/powershell/Private/Get-HubIdentifier.ps1:17]: I went ahead and opened the PR branch and just saved the files to auto-apply the current settings. There are a set of extensions that do this automatically. I think the PowerShell extension is the one working here. I just realized that isn't configured as a recommended extension, tho. I'll add that to this PR.
image

**Comment by jamelachahbar** [comment]
@flanakin , do you want me to squash and merge this?

**Comment by flanakin** [comment]
@jamelachahbar I looked at the integration tests we have here, which weren't working before. I was able to get something working for now, so we have an automated test for this. I updated that in the branch, so we're in a good state. 

Thanks for getting this in!
